### PR TITLE
ENH: Add capability to replace image

### LINKED ISSUE #1546: Enable image compression
## Explanation

I want to replace images in a pdf with compressed ones.
Gettings the images and saving them to disk work like a charm with the example in doc.
But I cannot change them in the pdf

## Code Example

How would your feature be used?

```python
from pypdf import PdfReader, PdfWriter
reader = PdfReader(input_file_path, strict=False)
for page in reader.pages:
        page.images = [compress(image_file_object) for image_file_object in page.images]
        writer.add_page(page)
...  # your new feature in action!
```

I have found a bunch of code that aimed at coding this feature:
- https://arunmozhi.in/2019/01/29/replacing-image-in-a-pdf-with-python/
- https://pypi.org/project/pdfminify/

but it doesn't work as expected.

### PR DESCRIPTION
**Pull request by pubpub-zz**
add capability to replace images.
This allow capabililty to compress files reducing image quality
closes #1546

## For the Reviewer

* Merge https://github.com/py-pdf/pypdf/pull/1848 first

### CODE REVIEW

**Comment by pubpub-zz** [comment]
this PR consists in PR #1848 + new commit : https://github.com/py-pdf/pypdf/pull/1849/commits/2d531d09a8ce3389f1a9dc53ca0c587087e6427f

File: pypdf/_utils.py [hunk 1]
```diff
@@ -508,3 +509,42 @@ class ImageFile(File):
 
     image: Optional[Any] = None  # optional ; direct PIL image access
     indirect_reference: Optional[IndirectObject] = None  # optional ; link to PdfObject
+
+    def replace(self, new_image: Any, **kwargs: Any) -> None:
+        """
+        replace the Image with a new PIL image
+        This is not allowed for inline image or image in a PdfReader
+        kwargs allows to pass parameters to `Image.Image.save()` such as quality
```
Comment (MartinThoma) [inline pypdf/_utils.py:517]:
```suggestion
        Replace the Image with a new PIL image.

    Args:
        new_image (Image.Image): The new PIL image to replace the existing image.
        **kwargs: Additional keyword arguments to pass to `Image.Image.save()`.

    Raises:
        TypeError: If the image is inline or in a PdfReader.
        TypeError: If the image does not belong to a PdfWriter.
        TypeError: If `new_image` is not a PIL Image.

    Note:
        This method replaces the existing image with a new image. It is not allowed for inline images or images within a PdfReader.
        The `kwargs` parameter allows passing additional parameters to `Image.Image.save()`, such as quality.
    """
```

File: pypdf/_utils.py [hunk 2]
```diff
@@ -508,3 +509,53 @@ class ImageFile(File):
...
+            **kwargs: Additional keyword arguments to pass to `Image.Image.save()`.
+
+        Raises:
+            TypeError: If the image is inline or in a PdfReader.
+            TypeError: If the image does not belong to a PdfWriter.
+            TypeError: If `new_image` is not a PIL Image.
+
+        Note:
+            This method replaces the existing image with a new image. It is not allowed for inline images or images within a PdfReader.
+            The `kwargs` parameter allows passing additional parameters to `Image.Image.save()`, such as quality.
```
Comment (MartinThoma) [inline pypdf/_utils.py:528]:
```suggestion
            This method replaces the existing image with a new image.
            It is not allowed for inline images or images within a PdfReader.
            The `kwargs` parameter allows passing additional parameters
            to `Image.Image.save()`, such as quality.
```

**Comment by MartinThoma** [comment]
Awesome work @pubpub-zz :partying_face: :clap:  Thank you :pray:

**Comment by finevine** [comment]
Thank you so much @pubpub-zz !!

**Pull request merged**